lightpath/controller.py

```diff
@@ -42,115 +50,403 @@ def __init__(self, client, endstations=None):
+        In the case of multiple valid paths, returns the path with latest
+        blocking device (highest blocking z-position)
 
         Parameters
         ----------
         device : Device
             A device somewhere in LCLS
 
+        Raises
+        ------
+        PathError
+            If a single, valid path cannot be determined
+
         Returns
         -------
         path : :class:`BeamPath`
             Path to and including given device
         """
-        try:
-            bl = self.beamlines[device.md.beamline]
-            prior, after = bl.split(device=device)
+        paths = list()
+        for src in self.sources:
+            if nx.has_path(self.graph, src, device.md.name):
+                found_paths = nx.all_simple_paths(self.graph, source=src,
+                                                  target=device.md.name)
+                paths.extend(found_paths)
+            else:
+                logger.debug(f'No path between {src} and {device.md.name}')
 
-        except KeyError:
-            raise ValueError("Beamline {} not found".format(bl.name))
+        if not paths:
+            raise PathError(f'No paths from sources to {device.md.name}')
+
+        subgraphs = [self.graph.subgraph(p) for p in paths]
+        beampaths = []
+        for subg in subgraphs:
+            devs = [node[1]['dev'] for node in subg.nodes.data()
+                    if node[1]['dev'] is not None]
+            beampaths.append(BeamPath(*devs, name=f'{device.md.name}_path'))
+
+        paths_by_length = sorted(beampaths, key=self.imped_z)
+
+        return paths_by_length[-1]
+
+    @staticmethod
+    def make_graph(
+        branch_devs: List[SearchResult],
+        branch_name: str,
+        sources: List[NodeName] = []
+    ) -> nx.DiGraph:
+        """
+        Create a graph with devices from branch_devs as nodes,
+        arranged in z-order.
+
+        It is assumed that all devices lie on branch: branch_name
+
+        If sources is provided, will prepend a source node at the
+        beginning of the branch when branch_name == sources
+
+        Parameters
+        ----------
+        branch_devs : List[happi.SearchResult]
+            a list of devices to generate graph with
 
-        return prior
+        branch_name : str
+            branch name, used to label edges
+
+        sources : List[NodeName], optional
+            source node Names, by default []
+
+        Returns
+        -------
+        nx.DiGraph
+            The branch comprised of nodes holding devices from branch_devs
+        """
+        graph = nx.DiGraph()
+        result_list = list(branch_devs)
+        result_list.sort(key=lambda x: x.metadata['z'])
+        # label nodes with device name, store ophyd device
+        nodes = []
+        for res in result_list:
+            try:
+                dev = res.get()
+            except Exception:
+                # TODO: be better about specific exceptions
```

Comment:
This one deserves some more thought for sure.  The knockon effects are pretty significant, since this would break the facility graph and path-finding methods.  

The dummy device that we'd have to make would also have to give a best effort guess at the input/output branches.  I might punt this to another PR, but will be necessary